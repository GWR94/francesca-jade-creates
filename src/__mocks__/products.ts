export const cakeProduct = {
  id: "8c01cf0d-9de5-4168-8143-73e1ea067ac9",
  title: "Racing Cake",
  description: "Test test test test test test",
  image: [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357819265-racecarcrop.jpg",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357839998-IMG_3565.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357864535-IMG_3659.JPG",
    },
  ],
  price: 0,
  shippingCost: 0,
  type: "Cake",
  tags: ["Cake", "Racing"],
  createdAt: "2020-03-09T11:54:02.873Z",
  updatedAt: "2020-03-16T11:24:33.081Z",
  setPrice: false,
};

export const createsProduct = {
  id: "3e743155-95e2-4ae4-8d70-eb5cb2d0c086",
  title: "Name Frame",
  description: "test test test test test test test test test test",
  image: [
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
    },
    {
      bucket: "fjadecreatesmedia121747-production",
      region: "eu-west-2",
      key:
        "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106268144-IMG_0305.JPEG",
    },
  ],
  price: 14.5,
  shippingCost: 4.5,
  type: "Creates",
  tags: ["Personalised", "Name", "Frame"],
  createdAt: "2020-03-09T12:51:52.637Z",
  updatedAt: "2020-03-20T14:24:14.718Z",
  setPrice: true,
};
