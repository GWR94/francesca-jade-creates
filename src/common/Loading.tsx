import React from "react";
import { CircularProgress, makeStyles } from "@material-ui/core";

interface LoadingProps {
  size?: number;
  color?: "inherit" | "primary" | "secondary";
  small?: boolean;
}

/**
 * A component which shows a loading spinner to the user to signal that there is
 * some loading going on.
 * @param {number} size - The size of the loading spinner
 * @param {string} color - The colour of the loading spinner
 */
const Loading: React.FC<LoadingProps> = ({
  size = 100,
  color = "inherit",
  small = false,
}): JSX.Element => {
  const useStyles = makeStyles({
    loading: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      height: "calc(100vh - 60px)",
    },
    loadingSmall: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      padding: 60,
    },
    spinner: {
      margin: "0 auto",
      display: "block",
    },
  });
  const classes = useStyles();
  return (
    <div className={small ? classes.loadingSmall : classes.loading}>
      <CircularProgress size={size} color={color} className={classes.spinner} />
    </div>
  );
};

export default Loading;
