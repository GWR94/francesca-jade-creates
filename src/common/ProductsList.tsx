import React from "react";
import { Drawer, Grid } from "@material-ui/core";
import _ from "underscore";
import { API } from "aws-amplify";
import ProductCard from "./product/ProductCard";
import { ProductProps } from "./interfaces/Product.i";
import { searchProducts } from "../graphql/queries";
import Pagination from "./Pagination";
import SearchFilter from "./SearchFilter";
import {
  ProductListProps,
  ProductListState,
  FilterProps,
  ProductFilters,
} from "../pages/accounts/interfaces/ProductList.i";
import { openSnackbar } from "../utils/Notifier";
import { INTENT } from "../themes";

/**
 * TODO
 * [ ] Tab that opens filters on mobile instead of always showing
 */

/**
 * Component which allows the user to filter each of the products, and allow them
 * to sort based on their needs. Used in ProductTypePage and AccountsPage. The former
 * is used to show all of their relevant product types - i.e Cakes or Creations, whereas
 * the AccountsPage gives admins the ability to view, change, filter and delete all
 * of the products available on the site.
 */
class ProductsList extends React.Component<ProductListProps, ProductListState> {
  public readonly state: ProductListState = {
    page: 1,
    queryResults: null,
    sortMethod: "createdAt",
    filterOpen: false,
  };

  /**
   * A method which handle the users input search query from the dropdown filters,
   * which can be accessed by clicking the pink hamburger icon on the screen.
   * @param query - the input which the user would like to query
   * @param filters - the active filters which the user has selected to filter out
   * unwanted products.
   */
  public handleSearchQuery = async (
    query: string,
    filters: ProductFilters,
  ): Promise<void> => {
    // get the type from props, which should be "Cake" or "Creates"
    const { type } = this.props;
    // destructure the filters, so they can be placed in the filtering object.
    const { searchQuery, adminFilters, sortBy } = filters;
    /**
     * if there is no userQuery and no admin filters, then set queryResults to null
     * so the original products can be viewed instead of a blank screen.
     */
    if (!query && !adminFilters) return this.setState({ queryResults: null });
    /**
     * initialise the filtering variable as an empty object so the relevant filters
     * can be added later.
     */
    let filtering: FilterProps = {};
    /**
     * set the sortBy filter in to state as sortMethod - which would either be
     * createdAt or updatedAt, based on the users input.
     */
    this.setState({ sortMethod: sortBy });
    // only process the query if the user has put a query into the search bar.
    if (query.length) {
      /**
       * if the searchQuery filter is set to be "all", then add the correct
       * filters to the filtering object. The "all" tag allows the user to
       * find all products where the products tags, title, tagline or
       * description matches that of a product.
       */
      if (searchQuery === "all") {
        /**
         * To make sure all of the possible properties are collected, the "or"
         * array should place all of the values in it.
         */
        filtering = {
          or: [
            { tags: { matchPhrasePrefix: query } },
            { title: { matchPhrasePrefix: query } },
            { description: { matchPhrasePrefix: query } },
          ],
        };
      } else {
        /**
         * If the searchQuery is a specific type (e.g. title/tags) then that
         * should be placed in the filtering object without an "or", as the user
         * clearly only wants that type to be filtered.
         */
        filtering = {
          [searchQuery]: {
            matchPhrasePrefix: query,
          },
        };
      }
    }

    /**
     * If there are any admin filters that have been passed through props, then they
     * need to be added to the filtering object.
     */
    if (adminFilters) {
      // destructure the adminFilters so they can be accessed easily.
      const { cake, creates } = adminFilters;
      /**
       * if the cake boolean value is true and creates boolean is false, then you only
       * want to show cakes. This means that the filtering.and needs to be set to equal
       * (eq) "Cake". If the opposite is true then filtering.and needs to be set to equal
       * "Creates". If cake & creates are both true/false then the type needs to be set
       * to be the type, which will come from props.
       */
      if (cake && !creates) {
        filtering.and = [{ type: { eq: "Cake" } }];
      } else if (!cake && creates) {
        filtering.and = [{ type: { eq: "Creates" } }];
      }
    } else {
      filtering.and = [{ type: { eq: type } }];
    }

    /**
     * If the filtering object is empty, then queryResults needs to be returned and
     * set into state, which will stop the execution of the function.
     */
    if (Object.keys(filtering).length === 0 && filtering.constructor === Object) {
      return this.setState({ queryResults: null });
    }

    try {
      /**
       * If the filtering object is not empty then the products table on the database
       * needs to be searched with the correct filters.
       */
      const { data } = await API.graphql({
        query: searchProducts,
        variables: {
          filter: filtering,
          limit: 100,
        },
        // @ts-ignore
        authMode: "API_KEY",
      });
      /**
       * remove duplicate products from search results which may occur from the query
       * being in tags/description/tagline etc, and store it in the queryResults variable
       * which will be set into state.
       */
      const queryResults = _.uniq(data.searchProducts.items, (x: ProductProps) => x.id);
      return this.setState({ queryResults });
    } catch (err) {
      /**
       * If there are any errors present, then notify the user with the snackbar error
       * component.
       */
      openSnackbar({
        message: "Unable to retrieve products. Please try again.",
        severity: INTENT.Danger,
      });
    }
  };

  public render(): JSX.Element {
    const { queryResults, sortMethod, page, filterOpen } = this.state;
    const { products, admin, history, type } = this.props;

    const results = queryResults || products;
    let maxPages = -1;
    if (products) {
      maxPages = Math.ceil(products.length / 12);
    }
    const productRange = {
      min: page === 1 ? 0 : (page - 1) * 12,
      max: page === 1 ? 12 : (page - 1) * 12 + 12,
    };

    return (
      <>
        <div
          className="product-list__filter-container"
          onClick={(e): void => {
            e.stopPropagation();
            this.setState({ filterOpen: true });
          }}
          role="button"
          tabIndex={0}
        >
          <i className="fas fa-bars product-list__filter-icon animated infinite pulse" />
        </div>
        <Grid container spacing={4}>
          {results && results.length ? (
            results
              .sort((a: ProductProps, b: ProductProps) =>
                // @ts-ignore //FIXME
                a[sortMethod] < b[sortMethod] ? 1 : -1,
              )
              .slice(productRange.min, productRange.max)
              .map(
                (product: ProductProps): JSX.Element => (
                  <Grid item lg={4} sm={6} xs={12} key={product.id}>
                    <ProductCard admin={admin} product={product} history={history} />
                  </Grid>
                ),
              )
          ) : (
            <div className="product-list__no-result-container">No Results</div>
          )}
        </Grid>
        {results && results.length > 12 && (
          <Pagination
            maxPages={maxPages}
            setPage={(page): void => this.setState({ page })}
            page={page}
          />
        )}
        <Drawer
          open={filterOpen}
          anchor="top"
          ModalProps={{
            onBackdropClick: (): void => this.setState({ filterOpen: false }),
            disableScrollLock: true,
          }}
        >
          <SearchFilter
            setQuery={(query, filters): Promise<void> =>
              this.handleSearchQuery(query, filters)
            }
            admin={admin}
            type={type}
          />
        </Drawer>
      </>
    );
  }
}

export default ProductsList;
