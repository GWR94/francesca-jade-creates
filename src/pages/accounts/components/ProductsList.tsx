import React from "react";
import { Col, Row } from "reactstrap";
import { Drawer } from "@blueprintjs/core";
import { API } from "aws-amplify";
import ProductCard from "../../../common/product/ProductCard";
import { ProductProps } from "../../../common/interfaces/Product.i";
import { searchProducts } from "../../../graphql/queries";
import Pagination from "../../../common/Pagination";
import SearchFilter from "../../../common/SearchFilter";
import {
  ProductListProps,
  ProductListState,
  FilterProps,
} from "../interfaces/ProductList.i";

/**
 * TODO
 * [ ] Tab that opens filters on mobile instead of always showing
 */

class ProductsList extends React.Component<ProductListProps, ProductListState> {
  public readonly state: ProductListState = {
    page: 1,
    queryResults: this.props.products,
    sortMethod: "createdAt",
    filterOpen: false,
  };

  public handleSearchQuery = async (query, filters): Promise<void> => {
    const { type } = this.props;
    const { searchQuery, adminFilters, sortBy } = filters;
    if (!query && !adminFilters) return this.setState({ queryResults: null });
    let filtering: FilterProps = {};
    this.setState({ sortMethod: sortBy });
    if (searchQuery === "all") {
      if (query.length > 0) {
        filtering = {
          or: [
            { tags: { matchPhrasePrefix: query } },
            { title: { matchPhrasePrefix: query } },
            { description: { matchPhrasePrefix: query } },
          ],
        };
      }
    } else {
      filtering = {
        [searchQuery]: {
          matchPhrasePrefix: query,
        },
      };
    }

    if (adminFilters) {
      if (adminFilters === "cakes") {
        filtering.and = [{ type: { eq: "Cake" } }];
      } else if (adminFilters === "creates") {
        filtering.and = [{ type: { eq: "Creates" } }];
      }
    } else {
      filtering.and = [{ type: { eq: type } }];
    }

    if (Object.keys(filtering).length === 0 && filtering.constructor === Object) {
      return this.setState({ queryResults: null });
    }

    try {
      const { data } = await API.graphql({
        query: searchProducts,
        variables: {
          filter: filtering,
          limit: 100,
        },
        // @ts-ignore
        authMode: "API_KEY",
      });
      console.log(data.searchProducts);
      return this.setState({ queryResults: data.searchProducts.items });
    } catch (err) {
      return console.error(err);
    }
  };

  public render(): JSX.Element {
    const { queryResults, sortMethod, page, filterOpen } = this.state;
    const { products, admin, history } = this.props;

    const results = queryResults || products;

    const maxPages = Math.ceil(products.length / 12);
    const productRange = {
      min: page === 1 ? 0 : (page - 1) * 12,
      max: page === 1 ? 12 : (page - 1) * 12 + 12,
    };

    return (
      <>
        <div
          className="product-list__filter-container"
          onClick={(): void => this.setState({ filterOpen: true })}
          role="button"
          tabIndex={0}
        >
          <i className="fas fa-bars product-list__filter-icon animated infinite pulse" />
        </div>
        <Row style={{ padding: "0 20px" }}>
          {results.length ? (
            results
              .sort((a, b) => (a[sortMethod] < b[sortMethod] ? 1 : -1))
              .slice(productRange.min, productRange.max)
              .map(
                (product: ProductProps): JSX.Element => (
                  <Col
                    lg={4}
                    sm={6}
                    xs={12}
                    key={product.id}
                    style={{ marginBottom: "30px" }}
                  >
                    <ProductCard admin={admin} product={product} history={history} />
                  </Col>
                ),
              )
          ) : (
            <div className="product-list__no-result-container">No Results</div>
          )}
        </Row>
        {results.length > 12 && (
          <Pagination
            maxPages={maxPages}
            setPage={(page): void => this.setState({ page })}
            page={page}
          />
        )}
        <Drawer
          isOpen={filterOpen}
          hasBackdrop={false}
          onClose={(): void => this.setState({ filterOpen: false })}
          position="top"
          size="auto"
          style={{
            margin: "0 auto",
            maxWidth: "400px",
            borderBottomLeftRadius: "8px",
            borderBottomRightRadius: "8px",
          }}
        >
          <SearchFilter
            setQuery={(query, filters): Promise<void> =>
              this.handleSearchQuery(query, filters)
            }
            admin={admin}
          />
        </Drawer>
      </>
    );
  }
}

export default ProductsList;
