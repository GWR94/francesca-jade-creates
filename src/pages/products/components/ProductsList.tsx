import React, { useEffect, useState } from "react";
import { CircularProgress, Grid, makeStyles, useMediaQuery } from "@material-ui/core";
import { useDispatch, useSelector } from "react-redux";
import { MenuRounded, Warning } from "@material-ui/icons";
import ProductCard from "../../../common/containers/ProductCard";
import { ProductProps } from "../../accounts/interfaces/Product.i";
import Pagination from "../../../common/Pagination";
import SearchFilter from "./SearchFilter";
import * as actions from "../../../actions/products.actions";
import { AppState } from "../../../store/store";
import { ProductListProps, ProductListState } from "../interfaces/ProductList.i";
import NonIdealState from "../../../common/containers/NonIdealState";
import styles from "../../accounts/styles/productList.style";
import { COLORS } from "../../../themes";

const initialState = {
  filterOpen: false,
  searchResults: null,
  isLoading: true,
  page: {
    min: 0,
    max: window.innerWidth > 600 ? 12 : 6,
  },
};
/**
 * Component which allows the user to filter each of the products, and allow them
 * to sort based on their needs. Used in ProductTypePage and AccountsPage. The former
 * is used to show all of their relevant product types - i.e Cakes or Creations, whereas
 * the AccountsPage gives admins the ability to view, change, filter and delete all
 * of the products available on the site.
 * @param type - The type of product that will be rendered - i.e Cake or Creates.
 * @param admin - Boolean value to determine if the user is an admin or not.
 */
const ProductsList: React.FC<ProductListProps> = ({
  type,
  admin,
  theme,
}): JSX.Element => {
  // create a variable for the useDispatch hook to be executed within the component
  const dispatch = useDispatch();
  // make styles using the styles and store it into a variable that can be executed.
  const useStyles = makeStyles(styles);
  // execute useStyles to create the classes object - which will contain all styles
  const classes = useStyles();
  // set initial state for component
  const [state, setState] = useState<ProductListState>(initialState);

  useEffect(() => {
    if (type) {
      /**
       * if there is a type prop that was passed down from the parent, dispatch the
       * getProducts action with the type passed as the parameter. This will return
       * all of the products which are of that type.
       */
      dispatch(
        actions.setSearchFilters({
          type: {
            eq: type,
          },
        }),
      );
      dispatch(actions.getProducts());
    }
  }, []);

  const desktop = useMediaQuery("(min-width: 600px)");

  // destructure relevant data from state
  const {
    filterOpen,
    searchResults,
    page: { min, max },
  } = state;
  /**
   * Retrieve the items from the products store by destructuring it and renaming
   * it to products for clarity's sake.
   */
  const { items: products, isSearching, noResults } = useSelector(
    ({ products }: AppState) => products,
  );
  /**
   * if there are any products (searchResults) passed from the SearchFilter component,
   * use them. If there aren't, then use the products from the products store (which
   * were retrieved from the useSelector hook).
   */
  const results = theme ? products || [] : searchResults || products;

  const isCake = type === "Cake";
  return (
    <>
      {!theme && (
        <div
          className={classes.filterContainer}
          onClick={(e): void => {
            e.stopPropagation();
            // open the search filter when clicking on the div
            setState({ ...state, filterOpen: true });
          }}
          role="button"
          tabIndex={0}
        >
          {/* Show an animated pulsing menu icon to the user */}
          <MenuRounded
            className={`${classes.filterIcon} animate__animated animate__infinite animate__pulse`}
          />
        </div>
      )}
      {!noResults ? (
        isSearching ? (
          <div
            style={{
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
              height: desktop ? 600 : 300,
              width: "100%",
            }}
          >
            <CircularProgress
              size={desktop ? 100 : 50}
              style={{ color: isCake ? COLORS.LightPink : COLORS.LightGray }}
            />
          </div>
        ) : (
          <Grid container spacing={2}>
            {/* 
            Map the results to a Grid item to control the responsive layout,
            and render a ProductCard component with the mapped product data
            passed through as props (and admin if the user is an admin).
          */}
            {results.slice(min, max).map(
              (product: ProductProps): JSX.Element => (
                <Grid item md={4} sm={6} xs={12} key={product.id}>
                  <ProductCard product={product} />
                </Grid>
              ),
            )}
          </Grid>
        )
      ) : (
        // If there are no products, show the NonIdealState component to notify the user
        <NonIdealState
          title="No Results Found."
          Icon={<Warning />}
          subtext="Please edit your search to return results."
        />
      )}
      {/* 
        If there are any results, and the size of array is 12 larger than 12, render 
        the Pagination component to control pages of products
      */}
      {results && results.length > 12 && (
        <Pagination
          dataLength={results.length}
          numPerPage={12}
          setPageValues={({ min, max }): void =>
            setState({ ...state, page: { min, max } })
          }
        />
      )}
      {/* 
        Render the drawer component with the SearchFilter component inside if
        the filterOpen boolean is true 
      */}
      {!theme && (
        <SearchFilter
          admin={admin}
          type={type}
          filterOpen={filterOpen}
          closeDrawer={(): void => setState({ ...state, filterOpen: false })}
        />
      )}
    </>
  );
};

export default ProductsList;
