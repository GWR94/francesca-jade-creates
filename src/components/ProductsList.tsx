import React, { Component } from "react";
import { H3 } from "@blueprintjs/core";
import { Col, Row } from "reactstrap";
import { API, graphqlOperation } from "aws-amplify";
import { listProducts } from "../graphql/queries";
import Loading from "./Loading";
import Product from "./Product";

interface Props {
  products: any;
}
interface State {}

class ProductsList extends Component<Props, State> {
  private onDeleteProduct = async (): Promise<void> => {
    // do something
  };

  public render(): JSX.Element {
    const { products } = this.props;
    console.log(products);
    return (
      <div>
        <H3>Products</H3>
        <Row>
          {products ? (
            products.map(
              (product): JSX.Element => (
                <Col md={4} sm={6} key={product.id}>
                  <Product {...product} />
                </Col>
              ),
            )
          ) : (
            <Loading size={100} />
          )}
        </Row>
      </div>
    );
  }
}

export default ProductsList;
