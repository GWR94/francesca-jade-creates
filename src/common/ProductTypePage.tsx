import React, { Component } from "react";
import { API } from "aws-amplify";
import { History } from "history";
import { Container } from "reactstrap";
import { listProducts, searchProducts } from "../graphql/queries";
import Loading from "./Loading";
import { ProductProps } from "./interfaces/Product.i";
import ProductsList from "../pages/accounts/components/ProductsList";
import Filter from "./Filter";

interface ProductTypeState {
  isLoading: boolean;
  products: ProductProps[];
  queryResults: ProductProps[];
}

interface ProductTypeProps {
  admin?: boolean;
  history: History;
  type: string;
}

/**
 * TODO
 * [ ] Fix creates showing up in cakes etc.
 */

export default class ProductTypePage extends Component<
  ProductTypeProps,
  ProductTypeState
> {
  public readonly state: ProductTypeState = {
    isLoading: true,
    products: null,
    queryResults: [],
  };

  public componentDidMount(): void {
    this.handleGetProducts();
  }

  private handleSearchQuery = async (query, searchTerms): Promise<void> => {
    let filtering;
    if (searchTerms === "all") {
      filtering = {
        or: [
          { tags: { matchPhrasePrefix: query } },
          { title: { matchPhrasePrefix: query } },
          { description: { matchPhrasePrefix: query } },
        ],
      };
    } else {
      filtering = {
        [searchTerms]: {
          matchPhrasePrefix: query,
        },
      };
    }
    const { data } = await API.graphql({
      query: searchProducts,
      variables: {
        filter: filtering,
      },
      // @ts-ignore
      authMode: "API_KEY",
    });
    this.setState({ products: data.searchProducts.items });
  };

  private handleGetProducts = async (): Promise<void> => {
    const { type } = this.props;
    const filtering =
      type === "Cake"
        ? {
            type: {
              eq: "Cake",
            },
          }
        : {
            type: {
              ne: "Cake",
            },
          };
    const { data } = await API.graphql({
      query: listProducts,
      variables: {
        filter: filtering,
      },
      // @ts-ignore
      authMode: "API_KEY",
    });
    this.setState({ products: data.listProducts.items, isLoading: false });
  };

  public render(): JSX.Element {
    const { isLoading, products, queryResults } = this.state;
    const { admin, history, handleSearchQuery } = this.props;

    const results = queryResults.length ? queryResults : products;
    return isLoading ? (
      <Loading size={100} />
    ) : (
      <Container>
        <Filter
          handleSearchQuery={(query, filter): Promise<void> =>
            this.handleSearchQuery(query, filter)
          }
        />
        <ProductsList products={results} admin={admin} noTitle />
      </Container>
    );
  }
}
