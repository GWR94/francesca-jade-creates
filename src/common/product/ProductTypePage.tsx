import React, { Component } from "react";
import { API } from "aws-amplify";
import { History } from "history";
import { Container } from "reactstrap";
import { H3 } from "@blueprintjs/core";
import { listProducts } from "../../graphql/queries";
import Loading from "../Loading";
import { ProductProps } from "../interfaces/Product.i";
import ProductsList from "../../pages/accounts/components/ProductsList";

interface ProductTypeState {
  isLoading: boolean;
  products: ProductProps[];
  queryResults: ProductProps[];
}

interface ProductTypeProps {
  admin?: boolean;
  history: History;
  type: "Cake" | "Creates";
}

export default class ProductTypePage extends Component<
  ProductTypeProps,
  ProductTypeState
> {
  public readonly state: ProductTypeState = {
    isLoading: true,
    products: null,
    queryResults: null,
  };

  public componentDidMount(): void {
    this.handleGetProducts();
  }

  public handleGetProducts = async (): Promise<void> => {
    const { type } = this.props;
    const filtering = type && {
      type: {
        eq: type,
      },
    };

    const { data } = await API.graphql({
      query: listProducts,
      variables: {
        filter: filtering,
        limit: 100,
      },
      // @ts-ignore
      authMode: "API_KEY",
    });
    this.setState({ products: data.listProducts.items, isLoading: false });
  };

  public render(): JSX.Element {
    const { isLoading, products, queryResults } = this.state;
    const { type, history } = this.props;
    const results = queryResults || products;
    return isLoading ? (
      <Loading size={100} />
    ) : (
      <Container>
        <h3 className="product-type__title">{type === "Cake" ? "Cakes" : "Creations"}</h3>
        <p className="product-type__description">-- Placeholder --</p>
        <p className="product-type__filter">
          To filter the products please click the pink button on the left hand side, and
          filter the results to your preferences.
        </p>
        <ProductsList products={results} noTitle type={type} history={history} />
      </Container>
    );
  }
}
