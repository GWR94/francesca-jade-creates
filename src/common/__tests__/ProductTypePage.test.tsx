import React from "react";
import { shallow } from "enzyme";
import ProductTypePage from "../product/ProductTypePage";
import { history } from "../../routes/Router";
import { createsProduct, cakeProduct } from "../../__mocks__/products";
import ProductsList from "../ProductsList";
import Loading from "../Loading";

describe("<ProductTypePage />", () => {
  it("should render the loading page before fetching products in handleGetProducts", () => {
    const wrapper = shallow(<ProductTypePage history={history} type="Cake" />);
    expect(wrapper.find(Loading).length).toBe(1);
    expect(wrapper.find(Loading).prop("size")).toBe(100);
  });

  it("should render correctly when rendering type 'Cake'", () => {
    const wrapper = shallow<ProductTypePage>(
      <ProductTypePage history={history} type="Cake" />,
    );
    wrapper.instance().handleGetProducts = jest.fn();
    wrapper.update();
    wrapper.instance().handleGetProducts();
    expect(wrapper.instance().handleGetProducts).toHaveBeenCalledTimes(1);
    wrapper.setState({
      isLoading: false,
      products: [cakeProduct],
    });
    expect(wrapper.find("h3.product-type__title").text()).toBe("Cakes");
    expect(wrapper.find("p.product-type__filter").text()).toBe(
      "To filter the products please click the pink button on the left hand side, and filter the results to your preferences.",
    );
    expect(wrapper.find(ProductsList).prop("products")).toStrictEqual([cakeProduct]);
    expect(wrapper.find(ProductsList).prop("type")).toBe("Cake");
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correctly when rendering type 'Creates'", () => {
    const wrapper = shallow<ProductTypePage>(
      <ProductTypePage history={history} type="Creates" />,
    );
    wrapper.instance().handleGetProducts = jest.fn();
    wrapper.update();
    wrapper.instance().handleGetProducts();
    expect(wrapper.instance().handleGetProducts).toHaveBeenCalledTimes(1);
    wrapper.setState({
      isLoading: false,
      products: [createsProduct],
    });
    expect(wrapper.find("h3.product-type__title").text()).toBe("Creations");
    expect(wrapper.find("p.product-type__filter").text()).toBe(
      "To filter the products please click the pink button on the left hand side, and filter the results to your preferences.",
    );
    expect(wrapper.find(ProductsList).prop("products")).toStrictEqual([createsProduct]);
    expect(wrapper.find(ProductsList).prop("type")).toBe("Creates");
    expect(wrapper).toMatchSnapshot();
  });
});
