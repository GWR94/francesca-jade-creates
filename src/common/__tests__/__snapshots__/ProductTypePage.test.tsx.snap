// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTypePage /> should render correctly when rendering type 'Cake' 1`] = `
<Container
  tag="div"
>
  <h3
    className="product-type__title"
  >
    Cakes
  </h3>
  <p
    className="product-type__description"
  >
    -- Placeholder --
  </p>
  <p
    className="product-type__filter"
  >
    To filter the products please click the pink button on the left hand side, and filter the results to your preferences.
  </p>
  <ProductsList
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    noTitle={true}
    products={
      Array [
        Object {
          "createdAt": "2020-03-09T11:54:02.873Z",
          "description": "Test test test test test test",
          "id": "8c01cf0d-9de5-4168-8143-73e1ea067ac9",
          "image": Array [
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357819265-racecarcrop.jpg",
              "region": "eu-west-2",
            },
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357839998-IMG_3565.JPEG",
              "region": "eu-west-2",
            },
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357864535-IMG_3659.JPG",
              "region": "eu-west-2",
            },
          ],
          "price": 0,
          "setPrice": false,
          "shippingCost": 0,
          "tags": Array [
            "Cake",
            "Racing",
          ],
          "title": "Racing Cake",
          "type": "Cake",
          "updatedAt": "2020-03-16T11:24:33.081Z",
        },
      ]
    }
    type="Cake"
  />
</Container>
`;

exports[`<ProductTypePage /> should render correctly when rendering type 'Creates' 1`] = `
<Container
  tag="div"
>
  <h3
    className="product-type__title"
  >
    Creations
  </h3>
  <p
    className="product-type__description"
  >
    -- Placeholder --
  </p>
  <p
    className="product-type__filter"
  >
    To filter the products please click the pink button on the left hand side, and filter the results to your preferences.
  </p>
  <ProductsList
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    noTitle={true}
    products={
      Array [
        Object {
          "createdAt": "2020-03-09T12:51:52.637Z",
          "description": "test test test test test test test test test test",
          "id": "3e743155-95e2-4ae4-8d70-eb5cb2d0c086",
          "image": Array [
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
              "region": "eu-west-2",
            },
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG",
              "region": "eu-west-2",
            },
            Object {
              "bucket": "fjadecreatesmedia121747-production",
              "key": "/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106268144-IMG_0305.JPEG",
              "region": "eu-west-2",
            },
          ],
          "price": 14.5,
          "setPrice": true,
          "shippingCost": 4.5,
          "tags": Array [
            "Personalised",
            "Name",
            "Frame",
          ],
          "title": "Name Frame",
          "type": "Creates",
          "updatedAt": "2020-03-20T14:24:14.718Z",
        },
      ]
    }
    type="Creates"
  />
</Container>
`;
